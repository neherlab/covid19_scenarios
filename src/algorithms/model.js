const msPerDay = 1000 * 60 * 60 * 24

const monthToDay = m => {
  return m * 30 + 15
}

const jan2020 = new Date('2020-01-01')

/**
 *
 * @param {number} time - point in time, until which simulation runs, as epoch time
 * @param {number} avgInfectionRate
 * @param {number} peakMonth - counting number in range of 0-11
 * @param {number} seasonalForcing -  seasonal variation in transmission. Usually a decimal number, e. g. 0.2
 * @returns {number}
 */
export function infectionRate(time, avgInfectionRate, peakMonth, seasonalForcing) {
  // this is super hacky
  const phase = ((time - jan2020) / msPerDay / 365 - monthToDay(peakMonth) / 365) * 2 * Math.PI
  return avgInfectionRate * (1 + seasonalForcing * Math.cos(phase))
}

export function getPopulationParams(params, severity, ageCounts, containment) {
  const pop = { ...params }

  pop.timeDeltaDays = 0.25
  pop.timeDelta = msPerDay * pop.timeDeltaDays
  pop.numberStochasticRuns = params.numberStochasticRuns

  // Compute age-stratified parameters
  const total = severity.map(d => d.ageGroup).reduce((a,b)=>a+ageCounts[b], 0);
  // TODO: Make this a form-adjustable factor
  pop.ageDistribution = {}
  pop.infectionSeverityRatio = {}
  pop.infectionFatality = {}
  pop.infectionCritical = {}
  pop.recoveryRate = {}
  pop.hospitalizedRate = {}
  pop.dischargeRate = {}
  pop.criticalRate = {}
  pop.deathRate = {}
  pop.overflowDeathRate = {}
  pop.stabilizationRate = {}
  pop.isolatedFrac = {}
  pop.importsPerDay = {}
  pop.importsPerDay.total = params.importsPerDay

  let hospitalizedFrac = 0
  let criticalFracHospitalized = 0
  let fatalFracCritical = 0
  let avgIsolatedFrac = 0
  severity.forEach(d => {
    const freq = (1.0 * ageCounts[d.ageGroup]) / total
    pop.ageDistribution[d.ageGroup] = freq
    pop.infectionSeverityRatio[d.ageGroup] = (d.severe / 100) * (d.confirmed / 100)
    pop.infectionCritical[d.ageGroup] = pop.infectionSeverityRatio[d.ageGroup] * (d.critical / 100)
    pop.infectionFatality[d.ageGroup] = pop.infectionCritical[d.ageGroup] * (d.fatal / 100)

    const dHospital = pop.infectionSeverityRatio[d.ageGroup]
    const dCritical = d.critical / 100
    const dFatal = d.fatal / 100

    hospitalizedFrac += freq * dHospital
    criticalFracHospitalized += freq * dCritical
    fatalFracCritical += freq * dFatal
    avgIsolatedFrac += (freq * d.isolated) / 100

    // Age specific rates
    pop.isolatedFrac[d.ageGroup] = d.isolated / 100
    pop.recoveryRate[d.ageGroup] = (1 - dHospital) / pop.infectiousPeriod
    pop.hospitalizedRate[d.ageGroup] = dHospital / pop.infectiousPeriod
    pop.dischargeRate[d.ageGroup] = (1 - dCritical) / pop.lengthHospitalStay
    pop.criticalRate[d.ageGroup] = dCritical / pop.lengthHospitalStay
    pop.stabilizationRate[d.ageGroup] = (1 - dFatal) / pop.lengthICUStay
    pop.deathRate[d.ageGroup] = dFatal / pop.lengthICUStay
    pop.overflowDeathRate[d.ageGroup] = pop.overflowSeverity*pop.deathRate[d.ageGroup];
  })

  // Get import rates per age class (assume flat)
  const L = Object.keys(pop.recoveryRate).length
  Object.keys(pop.recoveryRate).forEach(k => {
    pop.importsPerDay[k] = params.importsPerDay / L
  })

  // Population average rates
  pop.recoveryRate.total = (1 - hospitalizedFrac) / pop.infectiousPeriod
  pop.hospitalizedRate.total = hospitalizedFrac / pop.infectiousPeriod
  pop.dischargeRate.total = (1 - criticalFracHospitalized) / pop.lengthHospitalStay
  pop.criticalRate.total = criticalFracHospitalized / pop.lengthHospitalStay
  pop.deathRate.total = fatalFracCritical / pop.lengthICUStay
  pop.stabilizationRate.total = (1 - fatalFracCritical) / pop.lengthICUStay
  pop.overflowDeathRate.total = pop.overflowSeverity*fatalFracCritical / pop.lengthICUStay
  pop.isolatedFrac.total = avgIsolatedFrac

  // Infectivity dynamics
  const avgInfectionRate = pop.r0 / pop.infectiousPeriod
  pop.infectionRate = time =>
    containment(time) * infectionRate(time, avgInfectionRate, pop.peakMonth, pop.seasonalForcing)

  return pop
}

export function initializePopulation(N, numCases, t0, ages) {
  // FIXME: Why it can be `undefined`? Can it also be `null`?
  if (ages === undefined) {
    const put = x => {
      return { total: x }
    }
    return {
      time: t0,
      susceptible: put(N - numCases),
      exposed: put(0),
      infectious: put(numCases),
      hospitalized: put(0),
      critical: put(0),
      discharged: put(0),
      recovered: put(0),
      intensive: put(0),
      dead: put(0),
    }
  }
  const Z = Object.values(ages).reduce((a, b) => a + b)
  const pop = {
    time: t0,
    susceptible: {},
    exposed: {},
    infectious: {},
    hospitalized: {},
    critical: {},
    overflow: {},
    discharged: {},
    intensive: {},
    recovered: {},
    dead: {},
  }
  // TODO: Ensure the sum is equal to N!
  Object.keys(ages).forEach((k, i) => {
    const n = Math.round((ages[k] / Z) * N)
    pop.susceptible[k] = n
    pop.exposed[k] = 0
    pop.infectious[k] = 0
    pop.hospitalized[k] = 0
    pop.critical[k] = 0
    pop.overflow[k] = 0
    pop.discharged[k] = 0
    pop.recovered[k] = 0
    pop.intensive[k] = 0
    pop.dead[k] = 0
    if (i === Math.round(Object.keys(ages).length / 2)) {
      pop.susceptible[k] -= numCases
      pop.infectious[k] = 0.3 * numCases
      pop.exposed[k] = 0.7 * numCases
    }
  })

  return pop
}

// NOTE: Assumes all subfields corresponding to populations have the same set of keys
export function evolve(pop, P, sample) {
  const sum = (dict) => { return Object.values(dict).reduce((a, b) => a + b, 0); }
  const fracInfected = sum(pop.infectious) / P.populationServed

  const newTime = pop.time + P.timeDelta
  const newPop = {
    time: newTime,
    susceptible: {},
    exposed: {},
    infectious: {},
    recovered: {},
    hospitalized: {},
    critical: {},
    overflow: {},
    discharged: {},
    intensive: {},
    dead: {},
  }

  const push = (sub, age, delta) => {
    newPop[sub][age] = pop[sub][age] + delta
  }

  const newCases        = {};
  const newInfectious   = {};
  const newRecovered    = {};
  const newHospitalized = {};
  const newDischarged   = {};
  const newCritical     = {};
  const newStabilized   = {};
  const newICUDead      = {};
  const newOverflowStabilized = {};
  const newOverflowDead = {};

  // Compute all fluxes (apart from overflow states) barring no hospital bed constraints
  const Keys = Object.keys(pop.infectious).sort();
  Keys.forEach(age => {
    newCases[age] =
      sample(P.importsPerDay[age] * P.timeDeltaDays) +
      sample(
        (1 - P.isolatedFrac[age]) * P.infectionRate(newTime) * pop.susceptible[age] * fracInfected * P.timeDeltaDays,
      )
    newInfectious[age]         = Math.min(pop.exposed[age], sample(pop.exposed[age] * P.timeDeltaDays / P.incubationTime))
    newRecovered[age]          = Math.min(pop.infectious[age], sample(pop.infectious[age] * P.timeDeltaDays * P.recoveryRate[age]))
    newHospitalized[age]       = Math.min(pop.infectious[age] - newRecovered[age], sample(pop.infectious[age] * P.timeDeltaDays * P.hospitalizedRate[age]))
    newDischarged[age]         = Math.min(pop.hospitalized[age], sample(pop.hospitalized[age] * P.timeDeltaDays * P.dischargeRate[age]))
    newCritical[age]           = Math.min(pop.hospitalized[age] - newDischarged[age], sample(pop.hospitalized[age] * P.timeDeltaDays * P.criticalRate[age]))
    newStabilized[age]         = Math.min(pop.critical[age], sample(pop.critical[age] * P.timeDeltaDays * P.stabilizationRate[age]))
    newICUDead[age]            = Math.min(pop.critical[age] - newStabilized[age], sample(pop.critical[age] * P.timeDeltaDays * P.deathRate[age]))
    newOverflowStabilized[age] = Math.min(pop.overflow[age], sample(pop.overflow[age] * P.timeDeltaDays * P.stabilizationRate[age]))
    newOverflowDead[age]       = Math.min(pop.overflow[age] - newOverflowStabilized[age], sample(pop.overflow[age] * P.timeDeltaDays * P.overflowDeathRate[age]))

    push('susceptible', age, -newCases[age])
    push('exposed', age, newCases[age] - newInfectious[age])
    push('infectious', age, newInfectious[age] - newRecovered[age] - newHospitalized[age])
    push('hospitalized', age, newHospitalized[age] + newStabilized[age] + newOverflowStabilized[age] - newDischarged[age] - newCritical[age])
    // Cumulative categories
    push('recovered', age, newRecovered[age] + newDischarged[age])
    push('intensive', age, newCritical[age])
    push('discharged', age, newDischarged[age])
    push('dead', age, newICUDead[age] + newOverflowDead[age])
  })

  // Move hospitalized patients according to constrained resources
  let freeICUBeds = P.ICUBeds - (sum(pop.critical) - sum(newStabilized) - sum(newICUDead));
  Keys.forEach(age => {
      if (freeICUBeds > newCritical[age]) {
          freeICUBeds -= newCritical[age];
          push('critical', age, newCritical[age] - newStabilized[age] - newICUDead[age])
          push('overflow', age,  - newOverflowDead[age] - newOverflowStabilized[age])
      } else if (freeICUBeds > 0) {
          let newOverflow = newCritical[age] - freeICUBeds;
          push('critical', age, freeICUBeds - newStabilized[age] - newICUDead[age])
          push('overflow', age, newOverflow - newOverflowDead[age] - newOverflowStabilized[age])
          freeICUBeds = 0;
      } else {
          push('critical', age, -newStabilized[age] - newICUDead[age])
          push('overflow', age, newCritical[age] - newOverflowDead[age] - newOverflowStabilized[age])
      }
  });

  // If any overflow patients are left AND there are free beds, move them back.
  // Again, move w/ lower age as priority.
  for (let i = 0; freeICUBeds > 0 && i < Keys.length; i++) {
      const age = Keys[i];
      if (newPop.overflow[age] < freeICUBeds) {
          newPop.critical[age] += newPop.overflow[age];
          freeICUBeds -= newPop.overflow[age];
          newPop.overflow[age] = 0;
      } else {
          newPop.critical[age] += freeICUBeds;
          newPop.overflow[age] -= freeICUBeds;
          freeICUBeds = 0;
      }
  }

  // NOTE: For debug purposes only.
  /*
  const popSum = sum(newPop.susceptible) + sum(newPop.exposed) + sum(newPop.infectious) + sum(newPop.recovered) + sum(newPop.hospitalized) + sum(newPop.critical) + sum(newPop.overflow) + sum(newPop.dead);
  console.log(math.abs(popSum - P.populationServed));
  */

  return newPop
}

export function collectTotals(trajectory) {
  // FIXME: parameter reassign
  trajectory.forEach(d => {
    Object.keys(d).forEach(k => {
      if (k === 'time' || 'total' in d[k]) {
        return
      }
      d[k].total = Object.values(d[k]).reduce((a, b) => a + b)
    })
  })

  return trajectory
}

export function exportSimulation(trajectory) {
  // Store parameter values

  // Down sample trajectory to once a day.
  // TODO: Make the down sampling interval a parameter

  const header = Object.keys(trajectory[0])
  const csv = [header.join('\t')]

  const pop = {}
  trajectory.forEach(d => {
    const t = new Date(d.time).toISOString().slice(0, 10)
    if (t in pop) {
      return
    } // skip if date is already in table
    pop[t] = true
    let buf = ''
    header.forEach(k => {
      if (k === 'time') {
        buf += `${t}`
      } else {
        buf += `\t${Math.round(d[k].total)}`
      }
    })
    csv.push(buf)
  })

  return csv.join('\n')
}
