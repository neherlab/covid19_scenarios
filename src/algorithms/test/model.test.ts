import { infectionRate } from '../model'

describe('model', () => {
  describe('infectionRate', () => {
    const dec2020 = new Date('2020-12-01').getTime()

    it('baseline', () => {
      expect(infectionRate(dec2020, 0.9, 3, 0.2)).toEqual(0.7768945041075702)
    })

    it('accounts for time correctly', () => {
      const jul2020 = new Date('2020-07-01').getTime()
      const feb2021 = new Date('2021-02-01').getTime()

      expect(infectionRate(jul2020, 0.4, 3, 0.2)).toEqual(0.4194279777676749)
      expect(infectionRate(feb2021, 1.5, 3, 0.2)).toEqual(1.5927050983124844)
    })

    it('accounts for avgInfectionRate correctly', () => {
      expect(infectionRate(dec2020, 0.4, 3, 0.2)).toEqual(0.3452864462700312)
      expect(infectionRate(dec2020, 1.5, 3, 0.2)).toEqual(1.294824173512617)
    })

    it('accounts for peakMonth correctly', () => {
      expect(infectionRate(dec2020, 0.9, 2, 0.2)).toEqual(0.8577915498773262)
      expect(infectionRate(dec2020, 0.9, 7, 0.2)).toEqual(0.842905568053961)
    })

    it('accounts for seasonalForcing correctly', () => {
      expect(infectionRate(dec2020, 0.9, 3, 0.1)).toEqual(0.8384472520537851)
      expect(infectionRate(dec2020, 0.9, 3, 0.4)).toEqual(0.6537890082151402)
    })
  })
})
