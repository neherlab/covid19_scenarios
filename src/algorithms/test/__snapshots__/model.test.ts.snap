// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model evolve produces the expected output in the default scenario after 5 evolutions 1`] = `
Object {
  "cumulative": Object {
    "critical": Object {
      "0-9": 0,
      "10-19": 0,
      "20-29": 0,
      "30-39": 0,
      "40-49": 0,
      "50-59": 0.02018646950213547,
      "60-69": 0,
      "70-79": 0,
      "80+": 0,
    },
    "fatality": Object {
      "0-9": 0,
      "10-19": 0,
      "20-29": 0,
      "30-39": 0,
      "40-49": 0,
      "50-59": 0.0001450844569155264,
      "60-69": 0,
      "70-79": 0,
      "80+": 0,
    },
    "hospitalized": Object {
      "0-9": 2.3528344669060415e-7,
      "10-19": 6.659848850038048e-7,
      "20-29": 0.0000016581738039417507,
      "30-39": 0.00000291613242333308,
      "40-49": 0.000007441030932358732,
      "50-59": 0.7064185978419073,
      "60-69": 0.00003897512375721331,
      "70-79": 0.00005599687897778693,
      "80+": 0.0000611421008553007,
    },
    "recovered": Object {
      "0-9": 0.0004703316099345177,
      "10-19": 0.00044332393845086614,
      "20-29": 0.0005510664275099751,
      "30-39": 0.0006451132949840181,
      "40-49": 0.0006126448800975356,
      "50-59": 27.610884724340792,
      "60-69": 0.0004806931930056309,
      "70-79": 0.0003439808280064055,
      "80+": 0.00018342630256590212,
    },
  },
  "current": Object {
    "critical": Object {
      "0-9": 0,
      "10-19": 0,
      "20-29": 0,
      "30-39": 0,
      "40-49": 0,
      "50-59": 0.019823758359846656,
      "60-69": 0,
      "70-79": 0,
      "80+": 0,
    },
    "exposed": Object {
      "0-9": Array [
        6.590001975717847,
        1.8985981171330066,
        0.30181602536614993,
      ],
      "10-19": Array [
        6.217771457685203,
        1.7913622099584854,
        0.28476940256132427,
      ],
      "20-29": Array [
        7.740682663997832,
        2.2300977698967963,
        0.3545124559693652,
      ],
      "30-39": Array [
        9.075498661730135,
        2.6146449724862797,
        0.4156415244494359,
      ],
      "40-49": Array [
        8.684128631245459,
        2.5018951622832812,
        0.3977183908975536,
      ],
      "50-59": Array [
        31.811065284155312,
        44.32146633752035,
        48.824328815521056,
      ],
      "60-69": Array [
        7.277704457615031,
        2.0967183569838057,
        0.33330996324315043,
      ],
      "70-79": Array [
        5.60134716009214,
        1.613776505524084,
        0.25653971160439676,
      ],
      "80+": Array [
        3.4247226060643854,
        0.9867126369752011,
        0.15685928250809378,
      ],
    },
    "infectious": Object {
      "0-9": 0.025390342892641898,
      "10-19": 0.023956315474252338,
      "20-29": 0.029823369140801827,
      "30-39": 0.03496578129273958,
      "40-49": 0.03345801842198601,
      "50-59": 72.84207071259213,
      "60-69": 0.028039733974134673,
      "70-79": 0.021581525387770334,
      "80+": 0.013196024632719325,
    },
    "overflow": Object {
      "0-9": 0,
      "10-19": 0,
      "20-29": 0,
      "30-39": 0,
      "40-49": 0,
      "50-59": 0,
      "60-69": 0,
      "70-79": 0,
      "80+": 0,
    },
    "severe": Object {
      "0-9": 2.3528344669060415e-7,
      "10-19": 6.659848850038048e-7,
      "20-29": 0.0000016581738039417507,
      "30-39": 0.00000291613242333308,
      "40-49": 0.000007441030932358732,
      "50-59": 0.6258903465187388,
      "60-69": 0.00003897512375721331,
      "70-79": 0.00005599687897778693,
      "80+": 0.0000611421008553007,
    },
    "susceptible": Object {
      "0-9": 19930.183722971997,
      "10-19": 18802.681696624397,
      "20-29": 23415.644331016396,
      "30-39": 27458.858601030617,
      "40-49": 26273.382179711243,
      "50-59": 29528.94432493654,
      "60-69": 22013.263707819864,
      "70-79": 16935.506355119684,
      "80+": 10342.418264881415,
    },
  },
  "time": 1583157600000,
}
`;

exports[`model getPopulationParams produces the expected output in the default scenario 1`] = `
Object {
  "ICUBeds": 80,
  "ageDistribution": Object {
    "0-9": 0.10225114395521838,
    "10-19": 0.0964647560575147,
    "20-29": 0.12013546056661978,
    "30-39": 0.14087582334618426,
    "40-49": 0.13479387083572222,
    "50-59": 0.15259164409491252,
    "60-69": 0.11293803472872646,
    "70-79": 0.08688929452955889,
    "80+": 0.053059971885542775,
  },
  "frac": Object {
    "critical": Object {
      "0-9": 0.000025,
      "10-19": 0.00015000000000000001,
      "20-29": 0.00030000000000000003,
      "30-39": 0.0006749999999999999,
      "40-49": 0.0024000000000000002,
      "50-59": 0.00625,
      "60-69": 0.02625,
      "70-79": 0.063,
      "80+": 0.1375,
    },
    "fatal": Object {
      "0-9": 0.0000075,
      "10-19": 0.000045,
      "20-29": 0.00009,
      "30-39": 0.00020249999999999996,
      "40-49": 0.00072,
      "50-59": 0.0025000000000000005,
      "60-69": 0.0105,
      "70-79": 0.0315,
      "80+": 0.06875,
    },
    "isolated": Object {
      "0-9": 0,
      "10-19": 0,
      "20-29": 0,
      "30-39": 0,
      "40-49": 0,
      "50-59": 0,
      "60-69": 0,
      "70-79": 0,
      "80+": 0,
      "total": 0,
    },
    "severe": Object {
      "0-9": 0.0005,
      "10-19": 0.0015,
      "20-29": 0.003,
      "30-39": 0.0045,
      "40-49": 0.012,
      "50-59": 0.025,
      "60-69": 0.075,
      "70-79": 0.13999999999999999,
      "80+": 0.25,
    },
  },
  "hospitalBeds": 698,
  "importsPerDay": Object {
    "0-9": 0.011111111111111112,
    "10-19": 0.011111111111111112,
    "20-29": 0.011111111111111112,
    "30-39": 0.011111111111111112,
    "40-49": 0.011111111111111112,
    "50-59": 0.011111111111111112,
    "60-69": 0.011111111111111112,
    "70-79": 0.011111111111111112,
    "80+": 0.011111111111111112,
    "total": 0.1,
  },
  "numberStochasticRuns": 0,
  "populationServed": 195000,
  "rate": Object {
    "critical": Object {
      "0-9": 0.0125,
      "10-19": 0.025,
      "20-29": 0.025,
      "30-39": 0.0375,
      "40-49": 0.05,
      "50-59": 0.0625,
      "60-69": 0.0875,
      "70-79": 0.1125,
      "80+": 0.1375,
      "total": 0.05520552919584472,
    },
    "discharge": Object {
      "0-9": 0.2375,
      "10-19": 0.225,
      "20-29": 0.225,
      "30-39": 0.2125,
      "40-49": 0.2,
      "50-59": 0.1875,
      "60-69": 0.1625,
      "70-79": 0.1375,
      "80+": 0.11249999999999999,
      "total": 0.1947944708041553,
    },
    "fatality": Object {
      "0-9": 0.02142857142857143,
      "10-19": 0.02142857142857143,
      "20-29": 0.02142857142857143,
      "30-39": 0.02142857142857143,
      "40-49": 0.02142857142857143,
      "50-59": 0.028571428571428574,
      "60-69": 0.028571428571428574,
      "70-79": 0.03571428571428571,
      "80+": 0.03571428571428571,
      "total": 0.025324487226098872,
    },
    "infection": [Function],
    "latency": 0.2,
    "overflowFatality": Object {
      "0-9": 0.04285714285714286,
      "10-19": 0.04285714285714286,
      "20-29": 0.04285714285714286,
      "30-39": 0.04285714285714286,
      "40-49": 0.04285714285714286,
      "50-59": 0.05714285714285715,
      "60-69": 0.05714285714285715,
      "70-79": 0.07142857142857142,
      "80+": 0.07142857142857142,
      "total": 0.050648974452197744,
    },
    "recovery": Object {
      "0-9": 0.33316666666666667,
      "10-19": 0.33283333333333337,
      "20-29": 0.3323333333333333,
      "30-39": 0.33183333333333337,
      "40-49": 0.3293333333333333,
      "50-59": 0.325,
      "60-69": 0.30833333333333335,
      "70-79": 0.2866666666666667,
      "80+": 0.25,
      "total": 0.31982588844819954,
    },
    "severe": Object {
      "0-9": 0.00016666666666666666,
      "10-19": 0.0005,
      "20-29": 0.001,
      "30-39": 0.0014999999999999998,
      "40-49": 0.004,
      "50-59": 0.008333333333333333,
      "60-69": 0.024999999999999998,
      "70-79": 0.04666666666666666,
      "80+": 0.08333333333333333,
      "total": 0.013507444885133826,
    },
    "stabilize": Object {
      "0-9": 0.049999999999999996,
      "10-19": 0.049999999999999996,
      "20-29": 0.049999999999999996,
      "30-39": 0.049999999999999996,
      "40-49": 0.049999999999999996,
      "50-59": 0.04285714285714286,
      "60-69": 0.04285714285714286,
      "70-79": 0.03571428571428571,
      "80+": 0.03571428571428571,
      "total": 0.04610408420247255,
    },
  },
  "timeDelta": 21600000,
  "timeDeltaDays": 0.25,
}
`;

exports[`model getPopulationParams produces the expected output in the default scenario 2`] = `1.0282794619390596`;

exports[`model getPopulationParams produces the expected output in the default scenario 3`] = `0.7815218936243379`;

exports[`model getPopulationParams produces the expected output in the default scenario 4`] = `0.6412716928585847`;

exports[`model getPopulationParams produces the expected output in the default scenario 5`] = `0.5119976633870834`;

exports[`model getPopulationParams produces the expected output in the default scenario 6`] = `0.4777650045046975`;

exports[`model getPopulationParams produces the expected output in the default scenario 7`] = `0.45115800505877796`;

exports[`model getPopulationParams produces the expected output in the default scenario 8`] = `0.43543682301649544`;

exports[`model getPopulationParams produces the expected output in the default scenario 9`] = `0.43252777618729166`;

exports[`model getPopulationParams produces the expected output in the default scenario 10`] = `0.44278731033335716`;

exports[`model getPopulationParams produces the expected output in the default scenario 11`] = `0.4649583238392415`;

exports[`model initializePopulation produces the expected output in the default scenario 1`] = `
Object {
  "0-9": 19939,
  "10-19": 18811,
  "20-29": 23426,
  "30-39": 27471,
  "40-49": 26285,
  "50-59": 29542,
  "60-69": 22023,
  "70-79": 16943,
  "80+": 10347,
}
`;
